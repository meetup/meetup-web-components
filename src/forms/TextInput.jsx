import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import Icon from '../media/Icon';
import { MEDIA_SIZES } from '../utils/designConstants';
import withErrorList from '../utils/components/withErrorList';

export const FIELD_WITH_ICON_CLASS = 'field--withIcon';

/**
 * @module TextInput
 */
export const TextInput = (props) => {

	const {
		name,
		value,
		label,
		labelClassName,
		className,
		children,
		error,
		placeholder,
		id,
		onChange,
		isSearch,
		maxLength,
		pattern,
		disabled,
		iconShape,
		helperText,
		required,
		requiredText,
		...other
	} = props;

	const classNames = {
		field: cx(
			'span--100',
			{
				'field--error': error,
				[FIELD_WITH_ICON_CLASS]: iconShape
			},
			className
		),
		label: cx(
			'label--field',
			{
				'label--disabled': disabled,
				'label--required': required,
				'flush--bottom': helperText
			},
			labelClassName
		),
		helperText: cx(
			'helperTextContainer',
			{ required, disabled }
		),
		icon: 'icon--field',
	};

	const iconProps = {
		shape: iconShape,
		size: props.iconSize || 'xs',
		className: classNames.icon,
	};

	const paddingSize = parseInt(MEDIA_SIZES[iconProps.size], 10) + 24; // #TODO :SDS: replace '32' with something like "$space * 1.5" from `swarm-constants`
	const inputStyles = iconShape && {
		paddingLeft: `${paddingSize}px`
	};

	// WC-158
	// Only add a `value` prop if it is defined.
	// Workaround for IE11 support (see ticket)
	if (value || typeof value === 'string') {
		other.value = value;
	}

	return (
		<div className="inputContainer">
			{label &&
				<label className={classNames.label} htmlFor={id} data-requiredtext={required && requiredText}>
					{label}
				</label>
			}
			{helperText &&
				<div className={classNames.helperText}>
					{helperText}
				</div>
			}
			<div style={{position: 'relative'}}>
				<input type={isSearch ? 'search' : 'text'}
					name={name}
					required={required}
					placeholder={placeholder}
					className={classNames.field}
					onChange={onChange}
					pattern={pattern}
					disabled={disabled}
					id={id}
					style={inputStyles}
					{...other}
				/>
				{iconShape &&
					<Icon {...iconProps} />
				}
			</div>
			{ maxLength && <p tabIndex="-1" className='text--tiny text--secondary align--right charCount'>{parseInt(maxLength - value.length)}</p> }
			{children}
		</div>
	);
};

TextInput.propTypes = {
	name: PropTypes.string.isRequired,
	error: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	]),
	id: PropTypes.string,
	maxLength: PropTypes.number,
	pattern: PropTypes.string,
	label: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	]),
	labelClassName: PropTypes.string,
	placeholder: PropTypes.string,
	isSearch: PropTypes.bool,
	onChange: PropTypes.func,
	disabled: PropTypes.bool,
	iconShape: PropTypes.string,
	iconSize: PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl']),
	helperText: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	]),
	required: PropTypes.bool,
	requiredText: (props) => (
		props.required && !props.requiredText &&
			new Error('Inputs with `required` prop must provide also provide a translated string for "required" in the `requiredText` prop')
	)

};

export default withErrorList(TextInput);
