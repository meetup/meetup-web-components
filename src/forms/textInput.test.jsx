import React from 'react';
import { TextInput, FIELD_WITH_ICON_CLASS } from './TextInput';
import { mount } from 'enzyme';

const LABEL_TEXT = 'Super Hero',
	VALUE = 'Batman',
	NAME_ATTR = 'superhero',
	MAX_LEN = 20;

const DEFAULT_PROPS = {
	label: LABEL_TEXT,
	id: NAME_ATTR,
	name: NAME_ATTR,
	maxLength: MAX_LEN,
	value: VALUE,
	onChange: jest.fn(e => true),
};

const renderComponent = (props = {}) =>
	mount(<TextInput {...DEFAULT_PROPS} {...props} />);

describe('TextInput', function() {
	it('renders a required HTML <input> with expected attributes for mock data', () => {
		expect(renderComponent()).toMatchSnapshot();
	});

	it('should pass along specific props to the input element', () => {
		const inputEl = renderComponent({
			isSearch: true,
			disabled: true,
			required: true,
		}).find('input');

		// should have a name prop
		expect(inputEl.prop('name')).toEqual(NAME_ATTR);

		// should have a value prop
		expect(inputEl.prop('value')).toEqual(VALUE);

		// should have type set to search
		expect(inputEl.prop('type')).toEqual('search');

		// should have disabled prop
		expect(inputEl.prop('disabled')).not.toBeNull();

		// should have required prop
		expect(inputEl.prop('required')).not.toBeNull();
	});

	xit('should call onChange `props` function when input is changed', () => {
		const inputEl = renderComponent().find('input');
		const mockEvent = {
			target: {
				value: `${VALUE}`,
				setCustomValidity: () => true,
			},
		};

		inputEl.simulate('change', mockEvent);

		expect(DEFAULT_PROPS.onChange).toHaveBeenCalledWith(mockEvent);
	});

	// TODO: THIS TEST FAILS FOR A REAL REASON!
	// WE NEED TO ADD SUPPORT FOR `iconShape` PROP TO THE SWARM-COMPONENTS TEXTINPUT
	it('should show an icon when iconShape prop is specified', () => {
		const inputEl = renderComponent({
			iconShape: 'search',
		}).find('input');
		expect(inputEl.find(`.${FIELD_WITH_ICON_CLASS}`).exists()).toBe(true);
	});

	it('should have a label when label is given', () => {
		const labelEl = renderComponent()
			.find('label')
			.getDOMNode();
		expect(labelEl.textContent).toEqual(LABEL_TEXT);
	});
});
