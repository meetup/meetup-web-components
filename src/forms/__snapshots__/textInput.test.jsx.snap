// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders a required HTML <input> with expected attributes for mock data 1`] = `
<div
  className="inputContainer"
>
  <label
    className="label--field label--required"
    htmlFor="superhero"
  >
    Super Hero
  </label>
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <input
      className="span--100"
      id="superhero"
      name="superhero"
      required={true}
      type="text"
      value="Batman"
    />
  </div>
</div>
`;
